c10/core/DispatchKeySet.h

```diff
@@ -107,65 +317,104 @@ class DispatchKeySet final {
   uint64_t raw_repr() {
     return repr_;
   }
-  // Return the type id in this set with the highest priority (i.e.,
-  // is the largest in the DispatchKey enum).  Intuitively, this
-  // type id is the one that should handle dispatch (assuming there
-  // aren't any further exclusions or inclusions).
+
+  DispatchKey highestFunctionalityKey() const {
+    auto functionality_idx = indexOfHighestBit();
+    // This means that none of the functionality bits were set.
+    if (functionality_idx < num_backends) return DispatchKey::Undefined;
+    // The first num_backend bits in the keyset don't correspond to real dispatch keys.
+    return static_cast<DispatchKey>(functionality_idx - num_backends);
+  }
+
+  // This is similar like toBackendBit(DispatchKey), but less restrictive.
+  // toBackendBit() errors out if the key that it was passed has no backend bits,
+  // which is useful for error checking.
+  // We need a version of that here that can also handle "fake" backends like FPGA,
```

Comment:
Is the idea that fake backends behave like non-per-backend functionality and so this function will never be called for them?

Reply:
Yep that's right. If you're using a fake / non-customizable backend, then `highestBackendKey()` will return `BackendComponent::InvalidBit`. And there are two places where that matters:

(1) the runtime dispatch key calculation: The invariant is that if tensor's `key_set()` has no backend bits set, then it also can't have any per-backend-functionality bits set. Which means that we'll always dispatch to a "normal" functionality key and the backend bits will be masked out / ignored.

(2) autograd kernel registration: The `getAutogradRelatedKeySetFromBackend()` will return `AutogradOther` for those backends.